File: app/src/main/java/org/libreoffice/kit/LibreOfficeKit.java
package org.libreoffice.kit;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.res.AssetManager;
import android.util.Log;

import java.nio.ByteBuffer;

public final class LibreOfficeKit
{
    private static String LOGTAG = LibreOfficeKit.class.getSimpleName();
    private static AssetManager mgr;

    private LibreOfficeKit() {
    }

    public static void initializeLibrary() {
    }

    private static native boolean initializeNative(String dataDir, String cacheDir, String apkFile, AssetManager mgr);

    public static native ByteBuffer getLibreOfficeKitHandle();

    public static native void putenv(String string);

    public static native void redirectStdio(boolean state);

    static boolean initializeDone = false;

    public static synchronized void init(Context context)
    {
        if (initializeDone) {
            return;
        }

        mgr = context.getResources().getAssets();

        ApplicationInfo applicationInfo = context.getApplicationInfo();
        String dataDir = applicationInfo.dataDir;
        Log.i(LOGTAG, String.format("Initializing LibreOfficeKit, dataDir=%s\n", dataDir));

        redirectStdio(true);

        String cacheDir = context.getCacheDir().getAbsolutePath();
        String apkFile = context.getPackageResourcePath();

        if (!initializeNative(dataDir, cacheDir, apkFile, mgr)) {
            Log.e(LOGTAG, "Initialize native failed!");
            return;
        }
        initializeDone = true;
    }

    public static void loadNativeLibraries(String libraryDir) {
        NativeLibLoader.setLibraryDir(libraryDir);
        NativeLibLoader.load();
    }

    static {
    }
}

class NativeLibLoader {
    private static boolean done = false;
    private static String libraryDir = null;

    public static void setLibraryDir(String dir) {
        libraryDir = dir;
    }

    protected static synchronized void load() {
        if (done)
            return;
        
        if (libraryDir != null) {
            System.load(libraryDir + "/libc++_shared.so");
            System.load(libraryDir + "/liblo-native-code.so");
        } else {
            System.loadLibrary("c++_shared");
            System.loadLibrary("lo-native-code");
        }
        done = true;
    }
}

File: app/src/main/kotlin/org/libreoffice/kit/LibreOfficeKit.kt
package org.libreoffice.kit

import android.content.Context
import android.content.res.AssetManager
import android.util.Log
import java.nio.ByteBuffer

class LibreOfficeKit private constructor() {

    companion object {
        private val LOGTAG: String = LibreOfficeKit::class.java.simpleName
        private var mgr: AssetManager? = null

        @JvmStatic
        fun initializeLibrary() {
        }

        @JvmStatic
        private external fun initializeNative(dataDir: String, cacheDir: String, apkFile: String, mgr: AssetManager): Boolean

        @JvmStatic
        external fun getLibreOfficeKitHandle(): ByteBuffer

        @JvmStatic
        external fun putenv(string: String)

        @JvmStatic
        external fun redirectStdio(state: Boolean)

        @JvmField
        internal var initializeDone: Boolean = false

        @JvmStatic
        @Synchronized
        fun init(context: Context) {
            if (initializeDone) {
                return
            }

            mgr = context.resources.assets

            val dataDir = context.filesDir.absolutePath
            Log.i(LOGTAG, String.format("Initializing LibreOfficeKit, dataDir=%s\n", dataDir))

            redirectStdio(true)

            val cacheDir = context.cacheDir.absolutePath
            val apkFile = context.packageResourcePath

            if (!initializeNative(dataDir, cacheDir, apkFile, mgr!!)) {
                Log.e(LOGTAG, "Initialize native failed!")
                return
            }
            initializeDone = true
        }

        @JvmStatic
        fun loadNativeLibraries(libraryDir: String) {
            NativeLibLoader.setLibraryDir(libraryDir)
            NativeLibLoader.load()
        }
    }
}

internal object NativeLibLoader {
    private var done = false
    private var libraryDir: String? = null

    fun setLibraryDir(dir: String) {
        libraryDir = dir
    }

    @Synchronized
    fun load() {
        if (done) return

        val dir = libraryDir
        if (dir != null) {
            System.load("$dir/libc++_shared.so")
            System.load("$dir/liblo-native-code.so")
        } else {
            System.loadLibrary("c++_shared")
            System.loadLibrary("lo-native-code")
        }
        done = true
    }
}
